extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeParsingHelper.java

```diff
@@ -162,55 +171,83 @@ public static JsonObject getInitialData(String html) throws ParsingException {
-                    String key = p.getString("key");
-                    if (key != null && key.equals("cver")) {
-                        return p.getString("value");
+    public static String getClientVersion() throws ParsingException {
+        if (clientVersion != null && !clientVersion.isEmpty()) return clientVersion;
+
+        // Test if hard-coded client version is valid
+        try {
+            final String url = "https://www.youtube.com/results?search_query=test&pbj=1";
+
+            Map<String, List<String>> headers = new HashMap<>();
+            headers.put("X-YouTube-Client-Name", Collections.singletonList("1"));
+            headers.put("X-YouTube-Client-Version",
+                    Collections.singletonList(HARDCODED_CLIENT_VERSION));
+            final String response = getDownloader().get(url, headers).responseBody();
+            if (response.length() > 50) { // ensure to have a valid response
+                clientVersion = HARDCODED_CLIENT_VERSION;
+                return clientVersion;
+            }
+        } catch (Exception ignored) {}
+
+        // Try extracting it from YouTube's website otherwise
+        try {
+            final String url = "https://www.youtube.com/results?search_query=test";
+            final String html = getDownloader().get(url).responseBody();
+            JsonObject initialData = getInitialData(html);
+            JsonArray serviceTrackingParams = initialData.getObject("responseContext").getArray("serviceTrackingParams");
+            String shortClientVersion = null;
+
+            // try to get version from initial data first
+            for (Object service : serviceTrackingParams) {
+                JsonObject s = (JsonObject) service;
+                if (s.getString("service").equals("CSI")) {
+                    JsonArray params = s.getArray("params");
+                    for (Object param : params) {
+                        JsonObject p = (JsonObject) param;
+                        String key = p.getString("key");
+                        if (key != null && key.equals("cver")) {
+                            clientVersion = p.getString("value");
+                            return clientVersion;
+                        }
                     }
-                }
-            } else if (s.getString("service").equals("ECATCHER")) {
-                // fallback to get a shortened client version which does not contain the last do digits
-                JsonArray params = s.getArray("params");
-                for (Object param: params) {
-                    JsonObject p = (JsonObject) param;
-                    String key = p.getString("key");
-                    if (key != null && key.equals("client.version")) {
-                        shortClientVersion = p.getString("value");
+                } else if (s.getString("service").equals("ECATCHER")) {
+                    // fallback to get a shortened client version which does not contain the last do digits
```

Comment:
Did you mean two instead of do?

Reply:
[Blame](https://github.com/TeamNewPipe/NewPipeExtractor/blame/dev/extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeParsingHelper.java#L188) @TobiGr ;)